Data/IDProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Diagnostics;
+using Anvil.CSharp.Logging;
+
+namespace Anvil.CSharp.Data
+{
+    /// <summary>
+    /// An instance that provides a unique IDs each time requested.
+    /// Includes a mechanism to detect when ID supply is near exhaustion.
+    /// </summary>
+    public class IDProvider
+    {
+        /// <summary>
+        /// Triggers the first time that any instance of <see cref="IDProvider"/> passes its
+        /// <see cref="SupplyWarningThreshold"/> for the first time.
+        /// This gives the application the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public static event Action<IDProvider> OnIDLimitGlobalWarning;
```

Comment:
Hmmmn... this seems like the kind of thing that I might do that you would then call me out on :D

I'm not sure this should be static. If you create an `IDProvider` somewhere you should listen to that instance and handle it there too rather than having a global handler for any or all `IDProvider`s. Especially if someone else (different assembly, different third party mod etc) is handling this event before you get to and doing something they shouldn't.

Reply:
Oh ok, I just saw the instance event below... I'm still not sure we want to expose this though. But maybe i'm missing a use case?

Reply:
This is here as a high level application catch all. Rather than having to dig up every instance used you can listen to one static event. The only solution we know of at the moment is a soft reset and `uint` is such a large ID pool that it's an infrequent case. I think most applications will just want to know, at a high level, is some ID pool running out? Reload the whole application.

If you are writing an application that wants to only handle high risk `IDProviders` or specific subsets then you can still do that with the instance level event.

As a developer do you really want to go through and figure out everywhere an `IDProvider` is used?
Or have to design every Anvil system that uses an `IDProvider` indicate this edge case?

A good example is `UpdateHandle`. Everytime we use `UpdateHandle` do we now have to listen for the event on each instance?

If someone used this event anywhere in Anvil I'd reject it. The idea is that one place in an application listens to this.

I'm open to other suggestions but to me this is the right balance of safety and effort vs likelihood.

Reply:
Right, ok that makes sense. 

I wonder if we should do the following then:

1. Get rid of the instance event, and have it be an optional `Action` in the constructor. 

You either say "yes, we can gracefully handle this, or we know what we have to do to handle this and we will. No one else should interfere" OR you say "we can't handle this, it's up the higher level to do so via the static event."

2. If the instance Action is not null, then fire that and DON'T fire the static event. Upon creation we've stated we can and will handle it. If the instance Action is null, then fire the static event. The upper level can handle it if it chooses to. 

I think this makes the API safer and better deals with the intent. In `UpdateHandle`, we can't gracefully handle it. It's a signal for the app to soft restart. The static event should fire and app devs can handle it. 

But if we end up having something in the future that we can handle, we'll know that at creation and have passed in the instance Action to do something to handle it. I wouldn't want the app to get that static event and decide to soft restart since we've already coded the handling.

Reply:
That's a good point. I don't love the idea of submitting a callback into `IDProvider` but we could accomplish the same with the event and args that allow a handler to mark the event as "handled".

That would also allow others know about the warning occurring without necessarily needing to handle it. Since exhausted ID supply handling will be expensive in most cases I can see applications wanting to know when the operation has occurred and logging an event to analytics.